src/types/array-node.ts

```diff
@@ -100,15 +100,23 @@ export class ArrayNode extends Node {
 }
 
 export function createArrayFactory<S, T extends S>(subtype: IModelFactory<S, T>): IModelFactory<S[], IObservableArray<T>> {
-    return createFactory(
+    let factory = createFactory(
         "array-factory",
-        ArrayNode,
-        {
-            subType: subtype,
-            isArrayFactory: true
-        } as IArrayFactoryConfig,
-        () => observable.shallowArray()
+        "array",
+        snapshot => (Array.isArray(snapshot) && snapshot.every(item => subtype.is(item))),
```

Comment:
@mweststrate Yes, I'd like to support that as well. The default type problem occurs also with simple unions of structs. So my guess is to throw if there is no guessed type. You can change easily that by providing at first a dispatch function (to manually detect the type), or in near future use withDefault(unionOf(arrayOf(X), arrayOf(Y)), arrayOf(X)) to default to an empty array of X.